app/prj.conf

```diff
@@ -56,3 +56,7 @@ CONFIG_SCHED_CPU_MASK_PIN_ONLY=y
 CONFIG_SYS_CLOCK_TICKS_PER_SEC=15000
 CONFIG_DAI=y
 CONFIG_HEAP_MEM_POOL_SIZE=2048
+
+CONFIG_LLEXT=y
+CONFIG_LLEXT_STORAGE_WRITABLE=y
+CONFIG_MODULES=y
```

Comment:
If we have 3 different approaches to loading modules, I wouldn't use such generic name. It cause misunderstanding.

Reply:
@pjdobrowolski it isn't a name that we choose, it's hard-coded in kconfig tools. We want to use `tristate` in module Kconfig and set it to "m" to build as a module. And for that you need `CONFIG_MODULES`. And I'd propose to use `tristate` for any modules - whichever implementation they use, but that's up to respective authors / maintainers.

Reply:
>  modules:
>        The Symbol instance for the modules symbol. Currently hardcoded to
>        MODULES, which is backwards compatible. Kconfiglib will warn if
>        'option modules' is set on some other symbol. Tell me if you need proper
>        'option modules' support.
> 
Can't we add another specific symbol for LLEXT?

Reply:
> Tell me if you need proper 'option modules' support.

Please source quotations. This one comes from  here:
https://github.com/ulfalizer/Kconfiglib/blame/061e71f7d78cb0/kconfiglib.py#L671

This comment is 7 years old. Zephyr modules are months old and still in the making. I guess this text should be updated.



The reason @teburd and @lyakh could implement Zephyr modules without making any `kconfiglib.py` change is because `kconfiglib.py` has stuck to bug-for-bug compatibility with the original Kconfig implementation in C. https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html
I suspect the `kconfiglib.py` test suite covered modules even when Zephyr didn't support them yet.

Diverging `kconfiglib.py` away from the original Linux kernel implementation could break the `kconfiglib.py` test suite and make them slightly incompatible with each other, require slightly different documentations, disrupt other Zephyr users besides SOF,...A potentially huge amount of hassle for a small naming gain.

BTW `kconfiglib.py` is also used outside Zephyr.

I agree calling one of SOF's "3 different approaches" with a generic "CONFIG_MODULES" sucks, but Zephyr seems to have only one approach so it has no reason to disambiguate. This is hopefully just an implementation detail that most users will not pay much attention to: Kconfig has a user interface that shows the `help` text.

Note we have "_HOST_" abbreviated to "_HST_" and register sets called `DfPMCCH`. This is not an excuse not to care about names but it put things in a bit of perspective :-)



In any case this sounds like a Zephyr topic, not an SOF topic?

Reply:
Thanks for the context, @marc-hb 

> > Tell me if you need proper 'option modules' support.
> 
> Please source quotations. This one comes from here: https://github.com/ulfalizer/Kconfiglib/blame/061e71f7d78cb0/kconfiglib.py#L671
> 
> This comment is 7 years old. Zephyr modules are months old and still in the making. I guess this text should be updated.

Just to clarify a bit - it isn't just about the text (documentation), it's in the kconfiglib implementation. And yes, your text below actually alludes to that too.

[snip]

> I agree calling one of SOF's "3 different approaches" with a generic "CONFIG_MODULES" sucks, but Zephyr seems to have only one approach so it has no reason to disambiguate. This is hopefully just an implementation detail that most users will not pay much attention to: Kconfig has a user interface that shows the `help` text.

This isn't an option for **one** of the 3 different approaches. Other approaches are welcome and encouraged to use it too! IMHO it's a simple and natural way to switch a module between a built-in and modular build, so, any Kconfig options, that want this functionality, are free to use this! E.g. if we modularise our Maths libraries. As for which of the "three approaches" is used - I'd propose to use other means to make those decisions, if we ever have modules, that can be built with more than one option. E.g. indeed with this PR the smart-amp-test module would be buildable in three ways - built-in, as a module using system services or as an LLEXT. Currently The "y" Kconfig option selects whether it's built-in, selecting "m" build it as an LLEXT **immediately** during the base-firmware build, and to build it as a system-services module you run a separate cmake command, which ignores this Kconfig option completely, so you can build such a module regardless of this Kconfig value.

> Note we have "_HOST_" abbreviated to "_HST_" and register sets called `DfPMCCH`. This is not an excuse not to care about names but it put things in a bit of perspective :-)
> 
> In any case this sounds like a Zephyr topic, not an SOF topic?

Reply:
Thanks @lyakh . I didn't expect system-services modules to ignore the Kconfig of the base firmware (sounds quite risky) but if they do then this entire discussion seems indeed irrelevant.